server/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java

```diff
@@ -0,0 +1,448 @@
+import com.metamx.common.IAE;
+import com.metamx.common.RetryUtils;
+import com.metamx.common.StringUtils;
+import com.metamx.common.lifecycle.LifecycleStart;
+import com.metamx.common.lifecycle.LifecycleStop;
+import com.metamx.common.logger.Logger;
+import io.druid.guice.JsonConfigProvider;
+import io.druid.guice.LazySingleton;
+import io.druid.guice.LifecycleModule;
+import io.druid.guice.ManageLifecycle;
+import io.druid.guice.annotations.Json;
+import io.druid.initialization.DruidModule;
+import io.druid.query.extraction.namespace.ExtractionNamespace;
+import io.druid.query.extraction.namespace.ExtractionNamespaceFunctionFactory;
+import io.druid.query.extraction.namespace.JDBCExtractionNamespace;
+import io.druid.query.extraction.namespace.URIExtractionNamespace;
+import io.druid.server.initialization.ZkPathsConfig;
+import io.druid.server.namespace.cache.NamespaceExtractionCacheManager;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.PathChildrenCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheEvent;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheListener;
+import org.apache.curator.utils.ZKPaths;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executors;
+
+/**
+ *
+ */
+public class NamespacedExtractionModule implements DruidModule
+{
+  private static final String NAMESPACE_ENABLED_KEY = "enabledNamespaces";
+  private static final Logger log = new Logger(NamespacedExtractionModule.class);
+  private static final String TYPE_PREFIX = "druid.query.extraction.namespace.cache";
+  private final ConcurrentMap<String, Function<String, String>> fnCache = new ConcurrentHashMap<>();
+
+  @Override
+  public List<? extends Module> getJacksonModules()
+  {
+    return ImmutableList.<Module>of(
+        new Module()
+        {
+          @Override
+          public String getModuleName()
+          {
+            return "DruidNamespacedExtractionModule";
+          }
+
+          @Override
+          public Version version()
+          {
+            return Version.unknownVersion();
+          }
+
+          @Override
+          public void setupModule(SetupContext context)
+          {
+            context.registerSubtypes(ExtractionNamespace.class);
+          }
+        }
+    );
+  }
+
+  @Override
+  public void configure(Binder binder)
+  {
+    JsonConfigProvider.bind(binder, TYPE_PREFIX, NamespaceExtractionCacheManager.class);
+
+    binder
+        .bind(ExtractionNamespaceFunctionFactory.class)
+        .annotatedWith(Names.named(JDBCExtractionNamespace.class.getCanonicalName()))
```

Comment:
As a rule, don't use Named annotations for anything other than dynamic annotations that are local to a module.  If you need an annotation on an object, it should be worth it to create the annotation class.  If it's not worth it to create the annotation class, then that is a likely indication that you should rethink how you are doing the bindings.